### PR TITLE
Onboard Argus ACL

### PR DESCRIPTION
**Pull request by DiogoFerrao**

### CODE REVIEW

File: stackit/internal/services/argus/instance/resource.go [hunk 1]
```diff
@@ -292,11 +314,38 @@ func (r *instanceResource) Create(ctx context.Context, req resource.CreateReques
...
+	aclList, err := r.client.ListACL(ctx, *instanceId, projectId).Execute()
+	if err != nil {
+		core.LogAndAddError(ctx, &resp.Diagnostics, "Error creating instance", fmt.Sprintf("Calling API for ACL data: %v", err))
+		return
+	}
+
+	// Map response body to schema
+	err = mapFields(ctx, waitResp, aclList, &model)
+	if err != nil {
+		core.LogAndAddError(ctx, &resp.Diagnostics, "Error creating instance", fmt.Sprintf("Processing ACL API payload: %v", err))
```
Comment (joaopalet) [inline stackit/internal/services/argus/instance/resource.go:345]: This should actually be "Processing API response", right?

But we have it like this in the whole repo, so we could change it later

File: stackit/internal/services/argus/instance/resource.go [hunk 2]
```diff
@@ -325,18 +374,26 @@ func (r *instanceResource) Read(ctx context.Context, req resource.ReadRequest, r
...
-	// Set refreshed model
+
+	// Set state to fully populated data
 	diags = resp.State.Set(ctx, model)
 	resp.Diagnostics.Append(diags...)
 	if resp.Diagnostics.HasError() {
 		return
 	}
+
 	tflog.Info(ctx, "Argus instance created")
```
Comment (joaopalet) [inline stackit/internal/services/argus/instance/resource.go:397]:
```suggestion
	tflog.Info(ctx, "Argus instance read")
```

File: stackit/internal/services/argus/instance/resource.go [hunk 3]
```diff
@@ -500,6 +593,37 @@ func mapFields(ctx context.Context, r *argus.GetInstanceResponse, model *Model)
...
+	if aclList.Acl == nil || len(*aclList.Acl) == 0 {
+		model.ACL = types.SetNull(types.StringType)
+		return nil
+	}
+
+	acl := []attr.Value{}
+	for _, cidr := range *aclList.Acl {
+		acl = append(acl, types.StringValue(cidr))
+	}
+	aclsList, diags := types.SetValue(types.StringType, acl)
```
Comment (joaopalet) [inline stackit/internal/services/argus/instance/resource.go:622]: Maybe we can find a more descriptive name for this `aclsList` variable. `aclTF`? Or `aclSet`?

File: stackit/internal/services/argus/argus_acc_test.go [hunk 4]
```diff
@@ -236,7 +310,16 @@ func TestAccResource(t *testing.T) {
...
+					instanceResource["acl-1-updated"],
+				)),
+					fmt.Sprintf("%s-new", instanceResource["name"]),
+					instanceResource["new_plan_name"],
+					"",
+					"true",
+				),
 				Check: resource.ComposeAggregateTestCheckFunc(
 					// Instance
 					resource.TestCheckResourceAttr("stackit_argus_instance.instance", "project_id", instanceResource["project_id"]),
```
Comment (joaopalet) [inline stackit/internal/services/argus/argus_acc_test.go:325]: We should also add checks here to test that the ACLs are set and updates correctly

File: stackit/internal/services/argus/instance/resource.go [hunk 5]
```diff
@@ -292,11 +314,38 @@ func (r *instanceResource) Create(ctx context.Context, req resource.CreateReques
...
+
+	// Create ACL
+	err = updateACL(ctx, projectId, *instanceId, acl, r.client)
+	if err != nil {
+		core.LogAndAddError(ctx, &resp.Diagnostics, "Error creating instance", fmt.Sprintf("Creating ACL: %v", err))
+		return
+	}
+	aclList, err := r.client.ListACL(ctx, *instanceId, projectId).Execute()
+	if err != nil {
+		core.LogAndAddError(ctx, &resp.Diagnostics, "Error creating instance", fmt.Sprintf("Calling API for ACL data: %v", err))
```
Comment (vicentepinto98) [inline stackit/internal/services/argus/instance/resource.go:338]:
```suggestion
		core.LogAndAddError(ctx, &resp.Diagnostics, "Error creating instance", fmt.Sprintf("Calling API to list ACL data: %v", err))
```

File: stackit/internal/services/argus/instance/resource.go [hunk 6]
```diff
@@ -325,19 +374,27 @@ func (r *instanceResource) Read(ctx context.Context, req resource.ReadRequest, r
...
+		core.LogAndAddError(ctx, &resp.Diagnostics, "Error reading instance", fmt.Sprintf("Calling API for ACL data: %v", err))
+		return
+	}
+
 	// Map response body to schema
-	err = mapFields(ctx, instanceResp, &model)
+	err = mapFields(ctx, instanceResp, aclList, &model)
 	if err != nil {
-		core.LogAndAddError(ctx, &resp.Diagnostics, "Error reading instance", fmt.Sprintf("Processing API payload: %v", err))
+		core.LogAndAddError(ctx, &resp.Diagnostics, "Error reading instance", fmt.Sprintf("Processing ACL API payload: %v", err))
```
Comment (vicentepinto98) [inline stackit/internal/services/argus/instance/resource.go:386]:
```suggestion
		core.LogAndAddError(ctx, &resp.Diagnostics, "Error reading instance", fmt.Sprintf("Processing API payload: %v", err))
```

You're mapping both instance and acl data, not just acl

File: stackit/internal/services/argus/instance/resource.go [hunk 7]
```diff
@@ -386,6 +452,33 @@ func (r *instanceResource) Update(ctx context.Context, req resource.UpdateReques
...
+
+	// Update ACL
+	err = updateACL(ctx, projectId, instanceId, acl, r.client)
+	if err != nil {
+		core.LogAndAddError(ctx, &resp.Diagnostics, "Error updating instance", fmt.Sprintf("Updating ACL: %v", err))
+		return
+	}
+	aclList, err := r.client.ListACL(ctx, instanceId, projectId).Execute()
+	if err != nil {
+		core.LogAndAddError(ctx, &resp.Diagnostics, "Error updating instance", fmt.Sprintf("Calling API for ACL data: %v", err))
```
Comment (vicentepinto98) [inline stackit/internal/services/argus/instance/resource.go:464]:
```suggestion
		core.LogAndAddError(ctx, &resp.Diagnostics, "Error updating instance", fmt.Sprintf("Calling API to list ACL data: %v", err))
```

File: stackit/internal/services/argus/instance/datasource.go [hunk 8]
```diff
@@ -216,7 +221,13 @@ func (d *instanceDataSource) Read(ctx context.Context, req datasource.ReadReques
 		return
 	}
 
-	err = mapFields(ctx, instanceResponse, &model)
+	aclList, err := d.client.ListACL(ctx, instanceId, projectId).Execute()
+	if err != nil {
+		core.LogAndAddError(ctx, &resp.Diagnostics, "Error reading instance", fmt.Sprintf("Calling API for ACL data: %v", err))
```
Comment (vicentepinto98) [inline stackit/internal/services/argus/instance/datasource.go:226]:
```suggestion
		core.LogAndAddError(ctx, &resp.Diagnostics, "Error reading instance", fmt.Sprintf("Calling API to list ACL data: %v", err))
```

**Review by joaopalet — DISMISSED** [review]

**Review by vicentepinto98 — COMMENTED** [review]
nitpicks

File: stackit/internal/services/argus/argus_acc_test.go [hunk 9]
```diff
@@ -39,8 +42,9 @@ var credentialResource = map[string]string{
 	"project_id": testutil.ProjectId,
 }
 
-func resourceConfig(instanceName, planName, target, saml2EnableUrlParameters string) string {
-	return fmt.Sprintf(`
+func resourceConfig(acl *string, instanceName, planName, target, saml2EnableUrlParameters string) string {
```
Comment (vicentepinto98) [inline stackit/internal/services/argus/argus_acc_test.go:45]: you set `acl` as a pointer but there is no test case where it is passed as `nil`, so it seems unnecessary complexity

Comment (DiogoFerrao) [reply stackit/internal/services/argus/argus_acc_test.go:45]: This a good point, but there should probably be an acceptance test to make sure it still works if no ACL is provided. My commit below adds this

**Review by vicentepinto98 — APPROVED** [review]

**Comment by DiogoFerrao** [comment]
Closes #136

**Pull request merged**